cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineRunner.java

```diff
@@ -83,13 +83,12 @@ public SparkStreamingPipelineRunner(JavaSparkExecutionContext sec, JavaStreaming
     DataTracer dataTracer = sec.getDataTracer(stageInfo.getName());
     StreamingContext sourceContext = new DefaultStreamingContext(stageInfo.getName(), sec, streamingContext);
     JavaDStream<Object> javaDStream = source.getStream(sourceContext);
-    CountingTranformFunction<Object> function = new CountingTranformFunction<>(stageInfo.getName(), sec.getMetrics(),
-                                                                               "records.out", dataTracer);
     if (dataTracer.isEnabled()) {
-      javaDStream = javaDStream.transform(new LimitingFunction<>(function, numOfRecordsPreview));
-    } else {
-      javaDStream = javaDStream.transform(function);
+      // need to call transform here instead of filter because filter will create new function for each input record
```

Comment:
it will create a new function for each RDD, which would limit each RDD but not the entire DStream.